### ISSUE TITLE
Improve handling of low-memory situations

### ISSUE THREAD

**Issue by drumm** [php-tuf/composer-integration#127]
Reported at https://www.drupal.org/project/infrastructure/issues/3477553#comment-15846839:

> I also had problems with installation and now get out of memory errors when running composer up in both ddev and on my MacOS host.
> `PHP Fatal error:  Allowed memory size of 1610612736 bytes exhausted (tried to allocate 20480 bytes) in /var/www/html/vendor/php-tuf/php-tuf/src/CanonicalJsonTrait.php on line 46`

1.6MB is not a high memory limit. Whatever is usually necessary would be good to have in a friendly error message before getting to a fatal error.

If there are opportunities to reduce memory use, that is good to implement, too.

**Comment by catch56** [comment]
I think that was a typo on the issue, it OOMed at 1.6gb if I'm reading correctly.

**Comment by drumm** [comment]
Correct, I totally misread the conversion. 1.6gb is excessive memory use, so something is likely a blocker client-side. Next step is probably some debugging to confirm the memory usage and start seeing where it's going.

**Comment by catch56** [comment]
I've opened https://github.com/php-tuf/php-tuf/pull/386 and https://github.com/php-tuf/php-tuf/pull/387 against php-tuf/php-tuf which between them appear to bring memory usage more or less down to vanilla composer requirements.

This can probably be marked as duplicate however manual testing of both MRs would be very welcome.

**Comment by catch56** [comment]
#387 was reverted. Same approach is in https://github.com/php-tuf/php-tuf/pull/390 now.

**Comment by catch56** [comment]
After those two MRs, there is still a lot of time spent in Updater::searchDelegatedRolesForTarget() and the methods it calls.

I started looking at having targetsMetadata->getDelegatedRoles() implement a generator instead of producing the whole array at once, however that didn't actually bring results yet. It would be great to get reviews of the two current MRs, which would make tracking down the remaining issues easier.
